sky/spot/controller.py

```diff
@@ -145,8 +145,9 @@ def _run(self):
                             'cluster is healthy. Try to recover the job '
                             '(the cluster will not be restarted).')
 
-            resources = list(self._task.resources)[0]
-            if resources.need_cleanup_after_preemption():
+            resources = self._strategy_executor.launched_resources
+            if (resources is not None and
+                    resources.need_cleanup_after_preemption()):
```

Comment:
How about using the `handle.launched_resources` instead of adding a new attribute to the `_strategy_executor`?
```suggestion
            if handle is not None:
                resources = handle.launched_resources
                assert resources is not None, handle
                if resources.need_cleanup_after_preemption():
```

Reply:
Yes I considered this. however, this can't work as explained in the post. `handle` returned by
```
backend_utils.refresh_cluster_status_handle(
                 self._cluster_name, force_refresh=True)
```
will become `None` when the spot VM is preempted.

Reply:
Correct me if I am wrong, but if the spot VM is preempted, and the `refresh_cluster_status_handle` should already clean up the cluster right?

Reply:
For TPU VM, when it's preempted, we'll map such state to `None` so `refresh_cluster_status_handle` will return `None`.

I feel like letting the executor storing the cluster handle should be reasonable right? or is there any downside?

Reply:
> For TPU VM, when it's preempted, we'll map such state to None so refresh_cluster_status_handle will remove the row in the table and return None.

If the `refresh_cluster_status_handle` removes the TPU VM entry from the table, it will also clean up the cluster, right? Why do we need to clean it up again here?

> I feel like letting the executor storing the cluster handle should be reasonable right? or is there any downside?

The currently implementation will keep two copies of the handle in both the executor and the cache table. It can be easy to have the two copies un-synced, leading to unexpected error.

Reply:
> If the refresh_cluster_status_handle removes the TPU VM entry from the table, it will also clean up the cluster, right? Why do we need to clean it up again here? That is to say, we only need to clean up the cluster in the controller when the cluster is not in PREEMPTED status.

Oh right, that's true. Sorry I didn't think through. Then basically we have to guarantee such semantic in the future. i.e., when `refresh_cluster_status_handle` returns `None` it means everything on the console is cleaned up. just a small worry that this semantic may not be that obvious from its name. But should be okay.

> The current implementation will keep two copies of the handle in both the executor and the cache table. I feel like it can be easy to have the two copies un-synced, leading to unexpected errors.

I see. that makes sense.

Reply:
Good point! I think an important guarantee the `refresh_cluster_status_handle` should provide (I think we do) is that whenever the `status` is return with `None` value (same as `handle`), the cluster should be terminated in the console. Otherwise, we will experience resource leakage.